From a2cf989520fe172d400a2b8c31ee76e91b76da6a Mon Sep 17 00:00:00 2001
From: "zandobersek@gmail.com"
 <zandobersek@gmail.com@268f45cc-cd09-0410-ab3c-d52691b4dbfc>
Date: Mon, 16 Nov 2020 10:06:59 +0000
Subject: [PATCH xserver] [CoordinatedGraphics] Adjust client resizing,
 render-scope notifying in ThreadedCompositor::renderLayerTree()
 https://bugs.webkit.org/show_bug.cgi?id=218701

Reviewed by Carlos Garcia Campos.

Rework the client operations in ThreadedCompositor::renderLayerTree() so
that client resizing is done before the will-render client notification
is dispatched. This ensures everything in the will-render-to-did-render
scope is done for a well-specified size from the client's point of view.

Similar to this change, the GL viewport update is split from the client
resize operation and moved into the aforementioned rendering scope. This
change is mostly cosmetic, but it's done to neatly package all the GL
operations into that scope.

* Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::renderLayerTree):


git-svn-id: http://svn.webkit.org/repository/webkit/trunk@269842 268f45cc-cd09-0410-ab3c-d52691b4dbfc
---
 Source/WebKit/ChangeLog                       | 20 +++++++++++++++++++
 .../threadedcompositor/ThreadedCompositor.cpp | 20 ++++++++++++-------
 2 files changed, 33 insertions(+), 7 deletions(-)

diff --git a/Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp b/Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp
index b7cb12a1f8bf..e8f49ab186fc 100644
--- a/Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp
+++ b/Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp
@@ -176,8 +176,6 @@ void ThreadedCompositor::renderLayerTree()
     if (!m_context || !m_context->makeContextCurrent())
         return;
 
-    m_client.willRenderFrame();
-
     // Retrieve the scene attributes in a thread-safe manner.
     WebCore::IntSize viewportSize;
     WebCore::IntPoint scrollPosition;
@@ -204,15 +202,23 @@ void ThreadedCompositor::renderLayerTree()
         m_attributes.needsResize = false;
     }
 
-    if (needsResize) {
-        m_client.resize(viewportSize);
-        glViewport(0, 0, viewportSize.width(), viewportSize.height());
-    }
-
     TransformationMatrix viewportTransform;
     viewportTransform.scale(scaleFactor);
     viewportTransform.translate(-scrollPosition.x(), -scrollPosition.y());
 
+    // Resize the client, if necessary, before the will-render-frame call is dispatched.
+    // GL viewport is updated separately, if necessary. This establishes sequencing where
+    // everything inside the will-render and did-render scope is done for a constant-sized scene,
+    // and similarly all GL operations are done inside that specific scope.
+
+    if (needsResize)
+        m_client.resize(viewportSize);
+
+    m_client.willRenderFrame();
+
+    if (needsResize)
+        glViewport(0, 0, viewportSize.width(), viewportSize.height());
+
     glClearColor(0, 0, 0, 0);
     glClear(GL_COLOR_BUFFER_BIT);
 
-- 
2.30.2

